wicketd/src/artifacts/update_plan.rs

```diff
@@ -0,0 +1,1016 @@
+                    KnownArtifactKind::Trampoline,
+                ),
+            )?,
+            host_phase_2_hash: self.host_phase_2_hash.ok_or(
+                RepositoryError::MissingArtifactKind(KnownArtifactKind::Host),
+            )?,
+            control_plane_hash: self.control_plane_hash.ok_or(
+                RepositoryError::MissingArtifactKind(
+                    KnownArtifactKind::ControlPlane,
+                ),
+            )?,
+        })
+    }
+}
+
+// This function takes and returns `id` to avoid an unnecessary clone; `id` will
+// be present in either the Ok tuple or the error.
+fn read_hubris_board_from_archive(
+    id: ArtifactId,
+    data: Vec<u8>,
+) -> Result<(ArtifactId, Board), RepositoryError> {
+    let archive = match RawHubrisArchive::from_vec(data).map_err(Box::new) {
+        Ok(archive) => archive,
+        Err(error) => {
+            return Err(RepositoryError::ParsingHubrisArchive { id, error });
+        }
+    };
+    let caboose = match archive.read_caboose().map_err(Box::new) {
+        Ok(caboose) => caboose,
+        Err(error) => {
+            return Err(RepositoryError::ReadHubrisCaboose { id, error });
+        }
+    };
+    let board = match caboose.board() {
+        Ok(board) => board,
+        Err(error) => {
+            return Err(RepositoryError::ReadHubrisCabooseBoard { id, error });
+        }
+    };
+    let board = match std::str::from_utf8(board) {
+        Ok(s) => s,
+        Err(_) => {
+            return Err(RepositoryError::ReadHubrisCabooseBoardUtf8(id));
+        }
+    };
+    Ok((id, Board(board.to_string())))
+}
+
+// Record an artifact in `by_id` and `by_hash`, or fail if either already has an
+// entry for this id/hash.
+fn record_extracted_artifact(
+    tuf_repo_artifact_id: ArtifactId,
+    by_id: &mut BTreeMap<ArtifactId, Vec<ArtifactHashId>>,
+    by_hash: &mut HashMap<ArtifactHashId, ExtractedArtifactDataHandle>,
+    data: ExtractedArtifactDataHandle,
+    data_kind: ArtifactKind,
+    log: &Logger,
+) -> Result<(), RepositoryError> {
+    use std::collections::hash_map::Entry;
+
+    let artifact_hash_id =
+        ArtifactHashId { kind: data_kind, hash: data.hash() };
+
+    let by_hash_slot = match by_hash.entry(artifact_hash_id) {
+        Entry::Occupied(slot) => {
+            return Err(RepositoryError::DuplicateHashEntry(
+                slot.key().clone(),
+            ));
+        }
+        Entry::Vacant(slot) => slot,
+    };
+
+    info!(
+        log, "added artifact";
+        "name" => %tuf_repo_artifact_id.name,
+        "kind" => %by_hash_slot.key().kind,
+        "version" => %tuf_repo_artifact_id.version,
+        "hash" => %by_hash_slot.key().hash,
+        "length" => data.file_size(),
+    );
+
+    by_id
+        .entry(tuf_repo_artifact_id)
+        .or_default()
+        .push(by_hash_slot.key().clone());
+    by_hash_slot.insert(data);
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Are these tests mostly copied over?

Reply:
Yes, with minor tweaks to match the structural changes.